### PR TITLE
sysparam fixes, tests, spi flash refactoring

### PR DESCRIPTION
**Pull request by sheinz**
This PR includes:
 * Sysparam fixes for thread safety by @ourairquality 
 * Sysparam unit tests
 * Fix for negative int8 values
 * Sysparam bools without heap memory allocation
 * SPI flash refactoring, so spiffs and sysparam use the same flash access
 * spiflash test for unaligned access
 * switch from hand-written memcpy to the newlib's as discussed in https://github.com/SuperHouse/esp-open-rtos/issues/174 for spiflash

This PR allows https://github.com/SuperHouse/esp-open-rtos/pull/208 to be merged. I tested it and it works.

### CODE REVIEW

File: core/sysparam.c [hunk 1]
```diff
@@ -33,11 +31,14 @@
...
 #define SCAN_BUFFER_SIZE 8 // words
 
-/* The size of the temporary buffer used for reading back and verifying data
- * written to flash.  Making this larger will make the write-and-verify
- * operation slightly faster, but will use more heap during writes
+/* The size in words of the buffer used for reading keys when searching for a
+ * match, for reading payloads to check if the value has changed, and reading
+ * back from the flash to verify writes. Will work well if big enough for
+ * commonly used keys, and must be at least one word. Stack allocated so not too
+ * large!
```
Comment (ourairquality) [inline core/sysparam.c]: Looks like it is statically allocated now, which seems ok, so perhaps this 'stack allocated' comment is now stale.

File: core/sysparam.c [hunk 2]
```diff
@@ -119,50 +122,28 @@ static struct {
...
-    CHECK_FLASH_OP(sdk_spi_flash_write(addr, (void*) data, data_size));
-    return SYSPARAM_OK;
-}
-
-static inline IRAM sysparam_status_t _do_verify(uint32_t addr, const void *data, void *buffer, size_t len) {
-    CHECK_FLASH_OP(sdk_spi_flash_read(addr, buffer, len));
-    if (memcmp(data, buffer, len)) {
-        return SYSPARAM_ERR_IO;
+static sysparam_status_t _write_and_verify(uint32_t addr, const void *data, size_t data_size) {
+    static uint8_t bounce[BOUNCE_BUFFER_SIZE];
```
Comment (ourairquality) [inline core/sysparam.c]: This looks much cleaner using the improved flash support, thanks. Could we move all the flash read/write paths across to the new code now and remove the sdk versions (in follow up work)?

File: core/sysparam.c (hunk 1 above)
Comment (sheinz) [inline core/sysparam.c]: Actually, I added static during debugging and just forgot to remove it... I'm not sure what is actually better for that buffer to be static or be in stack.
I would actually prefer it to be in stack as it was before. As there's another such buffer in function `_compare_payload` which is in stack and it's better to be consistent.

File: core/sysparam.c [hunk 3]
```diff
@@ -295,17 +279,39 @@ static sysparam_status_t _find_entry(struct sysparam_context *ctx, uint16_t matc
...
+
+    CHECK_FLASH_OP(spiflash_read(addr, buffer, buffer_size));
+
+    return SYSPARAM_OK;
+}
+
+static inline sysparam_status_t _compare_payload(struct sysparam_context *ctx, uint8_t *value, size_t size) {
+    debug(3, "compare payload (%d) @ 0x%08x", size, ctx->addr);
+    if (ctx->entry.len != size) return SYSPARAM_NOTFOUND;
+    uint32_t *bounce[BOUNCE_BUFFER_WORDS];
```
Comment (sheinz) [inline core/sysparam.c]: This seems like a typo. The asterisk is redundant. Luckily pointer and uint32_t are the same size, so it worked without errors.

File: core/sysparam.c (hunk 2 above)
Comment (sheinz) [inline core/sysparam.c]: I think yes. It was tested quite intensively within spiffs tests.

**Comment by kanflo** [comment]
I think this one has been dormant for too long. It looks good to me and can be merged if no-one has any different opinion.

**Pull request merged**